NPE on Theme after upgrade to 21 when parent or import theme not exists

mposolda opened the pull request:
closes #17313

<!---
Please read https://github.com/keycloak/keycloak/blob/main/CONTRIBUTING.md and follow these guidelines when contributing to Keycloak
-->

mposolda committed:
NPE on Theme after upgrade to 21 when parent or import theme not exists
closes #17313

mposolda commented:
Docs PR with the note in the "Upgrading guide" that custom themes using `keycloak` admin theme won't work: https://github.com/keycloak/keycloak-documentation/pull/1788

stianst commented on services/src/main/java/org/keycloak/services/resources/admin/info/ServerInfoAdminResource.java:
```diff
@@ -179,16 +179,18 @@ private void setThemes(ServerInfoRepresentation info) {
-                    Theme theme = session.theme().getTheme(name, type);
-                    ThemeInfoRepresentation ti = new ThemeInfoRepresentation();
-                    ti.setName(name);
+                    Theme theme = session.theme().getTheme(name, type, false);
+                    if (theme != null) {
```
What about doing something like this instead of adding the overloaded method to ThemeManager:

```suggestion
                    Theme theme = session.theme().getTheme(name, type);
                    if (theme != null && theme.getName().equals(name)) {
```

mposolda replied:
Updated as you suggested

stianst commented on server-spi/src/main/java/org/keycloak/models/ThemeManager.java:
```diff
@@ -26,6 +26,17 @@ public interface ThemeManager {
+     * @return
+     * @throws IOException
+     */
+    Theme getTheme(String name, Theme.Type type, boolean fallbackToDefaultTheme) throws IOException;
```
See below, but thinking we could do without this overloaded method, especially since it's only used in a single place.

stianst commented on services/src/main/java/org/keycloak/theme/DefaultThemeManager.java:
```diff
@@ -106,20 +115,19 @@ private Theme loadTheme(String name, Theme.Type type) {
-            String[] s = theme.getImportName().split("/");
-            themes.add(findTheme(s[1], Theme.Type.valueOf(s[0].toUpperCase())));
-        }
+        if (!processImportedTheme(themes, theme, name, type)) return null;
```
Shouldn't this return the default theme?

mposolda replied:
Actually not as `loadTheme` is private method, which is supposed to return the theme itself just if it exists and if it is OK. Even before this PR, the method returned null in case that theme itself is not found. So I've just extended this to return null also in case that "parent" is null or "import" is null (which means theme was found, but is incorrect).

The fallback to default theme is already done in the calling method https://github.com/keycloak/keycloak/blob/main/services/src/main/java/org/keycloak/theme/DefaultThemeManager.java#L68-L72 .

stianst commented on services/src/main/java/org/keycloak/theme/DefaultThemeManager.java:
```diff
@@ -106,20 +115,19 @@ private Theme loadTheme(String name, Theme.Type type) {
             for (String parentName = theme.getParentName(); parentName != null; parentName = theme.getParentName()) {
+                String currentThemeName = theme.getName();
                 theme = findTheme(parentName, type);
+                if (theme == null) {
```
Shouldn't this return the default theme?

mposolda replied:
See my comment https://github.com/keycloak/keycloak/pull/17350#discussion_r1121871082

mposolda committed:
Update per review

mposolda commented:
@stianst Can you please re-review? I've used separate commit if you want to see just the latest changes since last review (can squash it during merge).

I've updated to your suggestions - removed new method in `ThemeManager` and handled the case in the `ServerInfoAdminResource` itself.

Replied to your other comments.
